src/ByteCountingStream.php

```diff
@@ -0,0 +1,64 @@
+<?php
+namespace GuzzleHttp\Psr7;
+
+use Psr\Http\Message\StreamInterface;
+use SebastianBergmann\GlobalState\RuntimeException;
+
+/**
+ *  Stream decorator that counts the number of bytes read off of an underlying
+ *  read stream.
+ */
+class ByteCountingStream implements StreamInterface
+{
+    use StreamDecoratorTrait;
+
+    /** @var int Number of bytes allowed to read */
+    private $maximum;
+
+    /** @var int Number of bytes read off*/
+    private $bytesCount;
+    /**
+     * @param StreamInterface $stream Stream to wrap
+     * @param int             $bytes  Number of bytes to read
+     * @throws \InvalidArgumentException
+     */
+    public function __construct(StreamInterface $stream, $bytes)
+    {
+        $this->stream = $stream;
+
+        if (!is_int($bytes) || $bytes < 0) {
+            $msg = "Bytes to read should be non-negative integer, got {$bytes}.";
+            throw new \InvalidArgumentException($msg);
+        }
+
+        if ($bytes > $this->stream->getSize()) {
+            $msg = "Bytes to read should be less than or equal to stream size : {$this->stream->getSize()}, "
```

Comment:
This error message will likely be confusing when used in context. Let's change this to:

```
The ByteCountingStream decorator expects to be able to read X bytes from a stream, but the stream being decorated only contains Y bytes.
```

I think that will give much more information and context when this error pops up higher up the stack.